### ISSUE TITLE
Citi Credit Cards stopped working (again) - 403 Forbidden from Akamai

### ISSUE THREAD

**Issue by jantman** [captin411/ofxclient#58]
For me at least, Citi Credit Cards stopped working again on July 22 (5 days ago).

I actually called Citi and asked and was referred to their Online Technical Support at 1-866-544-5534. I called _them_ and was told (not sure how much I believe this) that it's a known issue with problems downloading transactions, they're working on it, and to use PDF instead. I explained that you can't import a PDF into finance software, and was told they have people working on it and to call back in 24 hours if it's not fixed.

The error I'm seeing is coming from Akamai, their CDN provider, so there's definitely either something broken, they shut Direct Connect off alltogether, or they're using Akamai to filter on user-agent or other headers or part of the OFX itself.

Here's the debug-level logging for what I'm seeing, with all sensitive information replaced with the string "REDACTED":

```
2018-07-27 17:39:59,999 [DEBUG client.py:154 - root._do_post() ] posting data to https://www.accountonline.com/cards/svc/CitiOfxManager.do
2018-07-27 17:40:00,005 [DEBUG client.py:174 - root._do_post() ] ---- request headers ----
2018-07-27 17:40:00,005 [DEBUG client.py:176 - root._do_post() ] Content-Type: application/x-ofx
2018-07-27 17:40:00,005 [DEBUG client.py:176 - root._do_post() ] Host: www.accountonline.com
2018-07-27 17:40:00,005 [DEBUG client.py:176 - root._do_post() ] Content-Length: 660
2018-07-27 17:40:00,005 [DEBUG client.py:176 - root._do_post() ] Connection: Keep-Alive
2018-07-27 17:40:00,006 [DEBUG client.py:176 - root._do_post() ] Accept: */*, application/x-ofx
2018-07-27 17:40:00,006 [DEBUG client.py:176 - root._do_post() ] User-Agent: httpclient
2018-07-27 17:40:00,006 [DEBUG client.py:178 - root._do_post() ] ---- request body (query) ----
2018-07-27 17:40:00,006 [DEBUG client.py:179 - root._do_post() ] OFXHEADER:100
DATA:OFXSGML
VERSION:102
SECURITY:NONE
ENCODING:USASCII
CHARSET:1252
COMPRESSION:NONE
OLDFILEUID:NONE
NEWFILEUID:2311c068b37b4efea03e18929b1d0c28

<OFX>
<SIGNONMSGSRQV1>
<SONRQ>
<DTCLIENT>20180727173959
<USERID>REDACTED
<USERPASS>REDACTED
<LANGUAGE>ENG
<FI>
<ORG>Citigroup
<FID>24909
</FI>
<APPID>QWIN
<APPVER>2500

</SONRQ>
</SIGNONMSGSRQV1>
<CREDITCARDMSGSRQV1>
<CCSTMTTRNRQ>
<TRNUID>REDACTED
<CLTCOOKIE>4
<CCSTMTRQ>
<CCACCTFROM>
<ACCTID>REDACTED
</CCACCTFROM>
<INCTRAN>
<DTSTART>20180627
<INCLUDE>Y
</INCTRAN>
</CCSTMTRQ>
</CCSTMTTRNRQ>
</CREDITCARDMSGSRQV1>
</OFX>
2018-07-27 17:40:00,083 [DEBUG client.py:183 - root._do_post() ] ---- response ----
2018-07-27 17:40:00,083 [DEBUG client.py:184 - root._do_post() ] {'fp': None, 'debuglevel': 0, '_method': 'POST', 'headers': <http.client.HTTPMessage object at 0x7f51d7b31550>, 'msg': <http.client.HTTPMessage object at 0x7f51d7b31550>, 'version': 11, 'status': 403, 'reason': 'Forbidden', 'chunked': False, 'chunk_left': 'UNKNOWN', 'length': 0, 'will_close': True, 'code': 403}
2018-07-27 17:40:00,083 [DEBUG client.py:185 - root._do_post() ] Headers: [('Server', 'AkamaiGHost'), ('Mime-Version', '1.0'), ('Content-Type', 'text/html'), ('Content-Length', '309'), ('Expires', 'Fri, 27 Jul 2018 21:40:00 GMT'), ('Date', 'Fri, 27 Jul 2018 21:40:00 GMT'), ('Connection', 'close')]
2018-07-27 17:40:00,083 [DEBUG client.py:186 - root._do_post() ] <HTML><HEAD>
<TITLE>Access Denied</TITLE>
</HEAD><BODY>
<H1>Access Denied</H1>
 
You don't have permission to access "http&#58;&#47;&#47;www&#46;accountonline&#46;com&#47;cards&#47;svc&#47;CitiOfxManager&#46;do" on this server.<P>
Reference&#32;&#35;18&#46;86c95640&#46;1532727600&#46;3a5b6dc
</BODY>
</HTML>
```

I'll keep an eye on this...

**Comment by jhaggard11** [comment]
Hmm, I am not having any issues.  Are you still getting this error?

**Comment by jantman** [comment]
Yup...

As shown above, I'm sending OFX version 102, APPID QWIN and APPVER 2500...

**Comment by jantman** [comment]
So I just called Citi back and actually got to speak to a supervisor. They _don't_ have any known issues impacting Direct Connect, just something with downloading via the UI. She reached out to someone who actually knows their systems and was told that there's no reason they can think of for the Akamai error on their end. Everyone was very nice, and apologetic that they couldn't really do any more troubleshooting for third-party software.

If nobody else is seeing this, I guess I'll close the issue... and maybe see what happens if I try the download from work, using a completely different network and hitting a different Akamai edge server.

**Comment by jantman** [comment]
Crap.

Ok, so I replicated the 403 from Akamai with `curl`. I then ran the same `curl` from an EC2 instance in AWS, and got my OFX back fine. Apparently their Akamai configs are either blocking my IP specifically, or blocking some netblock that I'm in, or... I don't know, or for some reason I'm hitting a broken Akamai edge node or something.

If I hit https://www.accountonline.com/ directly in a browser I get the redirect to https://www.citi.com so whatever's happening seems specific to the OFX endpoint. No idea why they'd block me, I've just been running the same OFX requests once a day for years.

I guess it's just me though. Closing this.
